driver/embedding_driver.hpp

```diff
@@ -0,0 +1,434 @@
+            weightGradHost[weight_grad_idx] +=
+                outputGrad[outGrad_tv.get_tensor_view_idx(outGrad_layout)] * scale;
+        }
+    }
+
+    return 0;
+}
+
+template <typename Tgpu, typename Tref>
+class EmbeddingDriver : public Driver
+{
+public:
+    EmbeddingDriver() : Driver()
+    {
+        miopenCreateTensorDescriptor(&inputTensor);
+        miopenCreateTensorDescriptor(&outputTensorGrad);
+        miopenCreateTensorDescriptor(&weightTensorGrad);
+
+        data_type = miopen_type<Tgpu>{};
+    }
+
+    std::vector<int> ComputeStrides(std::vector<int> input);
+    int AddCmdLineArgs() override;
+    int ParseCmdLineArgs(int argc, char* argv[]) override;
+    InputFlags& GetInputFlags() override { return inflags; }
+
+    int GetandSetData() override;
+    std::vector<int> GetInputTensorLengthsFromCmdLine();
+
+    int AllocateBuffersAndCopy() override;
+
+    int RunForwardGPU() override;
+    int RunForwardCPU(); // Verify implements it
+
+    int RunBackwardGPU() override;
+    int RunBackwardCPU(); // Verify implements it
+
+    Tref GetTolerance();
+
+    int VerifyBackward() override;
+    int VerifyForward() override;
+    ~EmbeddingDriver() override
+    {
+        miopenDestroyTensorDescriptor(inputTensor);
+        miopenDestroyTensorDescriptor(outputTensorGrad);
+        miopenDestroyTensorDescriptor(weightTensorGrad);
+    }
+
+private:
+    InputFlags inflags;
+
+    int forw;
+    bool isContiguous;
+
+    // Backwards
+    miopenTensorDescriptor_t inputTensor;
+    miopenTensorDescriptor_t outputTensorGrad;
+    miopenTensorDescriptor_t weightTensorGrad;
+
+    std::unique_ptr<GPUMem> in_dev;
+    std::unique_ptr<GPUMem> outGrad_dev;
+    std::unique_ptr<GPUMem> weightGrad_dev;
+
+    std::vector<int64_t> in;
+    std::vector<Tgpu> outGrad;
+    std::vector<Tgpu> weightGrad;
+    std::vector<Tref> weightGradHost;
+
+    std::vector<int32_t> indices_freq;
+    std::unique_ptr<GPUMem> indices_freq_dev;
+
+    int64_t padding_idx;
+    bool scale_grad_by_freq;
+};
+
+template <typename Tgpu, typename Tref>
+int EmbeddingDriver<Tgpu, Tref>::ParseCmdLineArgs(int argc, char* argv[])
+{
+    inflags.Parse(argc, argv);
+
+    if(inflags.GetValueInt("time") == 1)
+    {
+        miopenEnableProfiling(GetHandle(), true);
+    }
+
+    forw = inflags.GetValueInt("forw");
+
+    if(forw != 2)
+    {
+        MIOPEN_THROW("Invalid Forward|Backward Mode");
+    }
+
+    isContiguous       = inflags.GetValueInt("is_contiguous") == 1 ? true : false;
+    padding_idx        = inflags.GetValueInt("padding_idx");
+    scale_grad_by_freq = inflags.GetValueInt("scale_grad_by_freq") == 1 ? true : false;
+
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref>
+int EmbeddingDriver<Tgpu, Tref>::GetandSetData()
+{
+    std::vector<int> in_len = inflags.GetValueTensor("input_dims").lengths;
+    auto in_stride          = ComputeStrides(in_len);
+    SetTensorNd(inputTensor, in_len, in_stride, data_type);
```

Comment:
Do this instead
```suggestion
    if(SetTensorNd(inputTensor, in_len, in_stride, data_type) != miopenStatusSuccess)
        MIOPEN_THROW("Error parsing input tensor: " + inflags.GetValueStr("input_dims") + ".");
```
Fix other parts where you're using `SetTensorNd` as well

Reply:
In my opinion, this should be an optional choice, as in MIOpen, there is no standard for this type of error checking.

Reply:
I understand. Someone mentioned that the approach I provided is a common convention. However, I know that MIOpenDriver is primarily for testing MIOpen APIs, so I’m fine with you choosing your own coding style. I'll mark this as resolved.